sys/kern/vm_amap.c

```diff
@@ -0,0 +1,172 @@
+#define KL_LOG KL_VM
+#include <sys/klog.h>
+#include <sys/malloc.h>
+#include <sys/pool.h>
+#include <sys/pmap.h>
+#include <sys/vm_map.h>
+#include <sys/vm_amap.h>
+#include <sys/vm_physmem.h>
+#include <sys/refcnt.h>
+#include <sys/errno.h>
+
+/*
+ * The idea of amaps comes from "Design and Implementation of the UVM Virtual
+ * Memory System" by Charles D. Cranor. (https://chuck.cranor.org/p/diss.pdf)
+ *
+ * The current implementation is a simpler (and limited) version of UVM.
+ *
+ * Things that are missing now (compared to original UVM Memory System):
+ *  - COW support (Amap references pages directly. After COW is implemented amap
+ *    will use anons structures to reference pages.)
+ *  - vm_objects (Amaps describe virtual memory. To describe memory mapped files
+ *    or devices the vm_objects are needed.)
+ *
+ * Some limitations of current implementation:
+ *  - Amaps are not resizable. We allocate more memory for each amap
+ *    (adjustable with EXTRA_AMAP_SLOTS) to allow resizing of small amounts.
+ *  - Amap is managing a simple array of referenced pages so it may not be the
+ *    most effective implementation.
+ */
+
+/* Amap size will be increased by this number of slots to easier resizing. */
+#define EXTRA_AMAP_SLOTS 16
+
+struct vm_amap {
+  size_t slots;     /* Read-only. Do not modify. */
+  refcnt_t ref_cnt; /* Atomic. */
+  mtx_t mtx;        /* Mutex guarding page list and bitmap. */
+  vm_page_t **pg_list;
+  bitstr_t *pg_bitmap;
+};
+
+static void vm_amap_free(vm_amap_t *amap);
+
+static POOL_DEFINE(P_VM_AMAP_STRUCT, "vm_amap_struct", sizeof(vm_amap_t));
+static KMALLOC_DEFINE(M_AMAP, "amap_slots");
+
+int vm_amap_ref(vm_amap_t *amap) {
+  return amap->ref_cnt;
+}
+
+size_t vm_amap_slots(vm_amap_t *amap) {
+  return amap->slots;
+}
+
+vm_amap_t *vm_amap_alloc(size_t slots) {
+  slots += EXTRA_AMAP_SLOTS;
+  vm_amap_t *amap = pool_alloc(P_VM_AMAP_STRUCT, M_WAITOK);
```

Comment:
Oh. I have an interesting observation here. It appears that `pool_alloc` always has `M_WAITOK` semantics, which states it never returns NULL - so `nullptr` checks are unnecessary.

Reply:
Okay, I have removed all the NULL checks here.